Improve REPL result printing for String and Product [ci: last-only]

som-snytt opened the pull request:
Add surrounding quotes to Strings and use verbose toString for Products that are not Tuples.

The goal is limited pretty-printing in REPL.

When arity >= 2 and not opting out with deprecatedName, print param names.

The opt-out idiom is `@deprecatedName("x") x: X`, which means "don't call me by my name." If any parameter is annotated, just use ordinary toString.

Note: goal is not pretty-printing but to improve casual ~toString~ REPL experience.

If there is only one param, no disambiguation is required, and the name of the wrapped value is often uninteresting (supposition).

Fixes scala/bug#8603

som-snytt commented:
Check if param check is too forcing at time of invocation.

`_toString0` not yet enshrined.

sjrd commented:
That's a *massive* run-time behavior change for a lot of code. It's going to break so many things. I really don't we should do that in a minor version, and I would even be very hesitant to do it in a major version.

som-snytt commented:
@sjrd yes, you've registered that position in the previous PR. This mostly just for fun; given past discussion, it probably won't fly. But the problem space is small enough to revisit.

som-snytt commented on test/files/run/infix.scala:
```diff
@@ -1,4 +1,4 @@
-case class op(x: op, y: Int, z: Int) {
+case class op(@deprecatedName("x") x: op, y: Int, z: Int) {
```
Sample opt-out, where the output calls for compact representation.

som-snytt commented on test/files/run/t1373.scala:
```diff
@@ -1,6 +1,6 @@
 // Testing whether case class params come back in the right order.
 object Test extends App {
-  case class Foo(private val a: String, b: String, private val c: String, d: String, private val e: String)
+  case class Foo(@deprecatedName("a","") private val a: String, b: String, private val c: String, d: String, private val e: String)
```
Sample opt-out for stringly testing.

dwijnand commented:
> yes, you've registered that position in the previous PR.

Whereabouts?  Couldn't find anything.

dwijnand commented on src/library/scala/runtime/ScalaRunTime.scala:
```diff
@@ -155,8 +155,23 @@ object ScalaRunTime {
+    while (more) {
+      sb.append(nm.next()).append(" = ").append(java.util.Objects.toString(it.next()))
+      more = it.hasNext
+      if (more) sb.append(", ")
```
❤️

som-snytt commented:
Previous art and discussion
https://github.com/scala/scala/pull/6936
https://contributors.scala-lang.org/t/case-class-tostring-new-behavior-proposal-with-implementation/2056

joshlemer commented on src/library/scala/runtime/ScalaRunTime.scala:
```diff
@@ -155,8 +155,23 @@ object ScalaRunTime {
+    while (more) {
+      sb.append(nm.next()).append(" = ").append(java.util.Objects.toString(it.next()))
+      more = it.hasNext
+      if (more) sb.append(", ")
```
I recommend to avoid the spacers, either on the sides of `=` or after commas, or both. So just `Foo(a=true,b=-1)` rather than `Foo(a = true, b = -1)`. Can actually make a big difference for larger case classes

som-snytt replied:
Thanks for the feedback. It got a heart from dwijnand, who has a big heart. This style dilemma is the camel-back-breaking issue. I guessed that people like to write the apply with spaces.

Footnote: maybe compiler can infer from the source text how the user prefers to write it! Maybe for more than N params, output as ascii table!

SethTisue replied:
This is highly bike-sheddable. I like the spaces, since that's how it would appear in nearly every Scala codebase I've seen.

som-snytt replied:
That's two big hearts for.

som-snytt commented:
Closing for good, as there is flak just to explore the problem space.

som-snytt commented:
Reopening because before it was fixed you would see:
```
scala> List[String]()
res1: List[String] = List()

scala> List[String]("")
res2: List[String] = List()
```

som-snytt commented:
Was hoping to tweak the check files and get green checks. This is new to me, failing commit Nr 2:
`[error]   - root/testJarSize failed: The library jar is too big: 5864554 bytes.`

The last fails with
```
[error] Test scala.tools.nsc.interpreter.ScriptedTest.t7933$u0020multiple$u0020eval$u0020compiled$u0020script failed: javax.script.ScriptException: too many arguments (found 3, expected 2) for method replStringOf: (arg: Any, maxElements: Int): String in val i = new java.util.concurrent.atomic.AtomicInteger at line number 8 at column number 150, took 0.559 sec
```
which means it's using latest REPL and previous library. That should be a partest.

som-snytt commented on test/files/run/t8603.check:
```diff
@@ -0,0 +1,96 @@
+xyz
+"""
+
+scala> "abc\r\nxyz"
```
This will be different on windoze, so the test must be guarded.

som-snytt commented on test/files/run/t8603.check:
```diff
@@ -0,0 +1,96 @@
+
+scala> E(C(42, "abc\rxyz\n", 3.14), D(42, Array(3.14, "xyz")))
+val res20: E =
+E(C(42,abc
```
The format was lost on recursion; something broke it.

som-snytt commented:
Probably there is something interesting here that is useful to folks, but it falls under repl improvements. 2.12 repl doesn't merge forward to 2.13 which won't merge to 3.

som-snytt commented:
/rebuild

som-snytt commented:
My previous https://github.com/scala/scala/pull/8885#issuecomment-664673566 is belied by Scala 3's reflective invocation of the new API, which will work as soon as it lands.

dwijnand commented on test/junit/scala/runtime/ScalaRunTimeTest.scala:
```diff
@@ -29,11 +29,11 @@ class ScalaRunTimeTest {
     val iterable = Iterable("a", "", " c", null)
     assertEquals(s"""List("a", "", " c", null)""", stringOf(iterable))
-    assertEquals(s"""List("a", "")""", stringOf(iterable, 2))
+    //assertEquals(s"""List("a", "")""", stringOf(iterable, 2))
```
What happened to these two?

som-snytt replied:
The API change means 2 is strlen, not element count. There is a quirky or buggy handling of n < 3 which could be fixed. I think the original thought was n >= 3 because of appended `"..."`.

SethTisue commented:
I suggest that the fix for scala/bug#12337 be submitted as a separate PR.

SethTisue commented:
The string quoting part could also be considered separately.

As for the main change, `-Vcase` and such, I think it's probably out of scope for Scala 2.

dwijnand commented:
For the improved case class printing use case, I think it would be best if we can find a solution that truly is REPL only, i.e. doesn't affect the bytecode that's emitted (e.g. in SyntheticMethods).

But I want to land all these pieces, and that'll be easier if we do them one by one.

som-snytt commented:
I'm not likely to invest more time into this effort. There are 3 bits, "minimal pretty print from REPL", "also for case class", "limit the output". The "string limit" depends on the API, so that is not an independent change. Putting the "use REPL-style printing for case class" behind a flag is a reasonable ask (in lieu of making it the default, which people have asked for). This is all in lieu of a typeclass solution.

SethTisue commented:
Okay. Anyone else interested in submitting either or both of those separate PRs?

som-snytt commented:
WFH was a sick day, so I pushed a commit from March 1 which tweaked str len limit. Also tweaked `-Vcase` to use repl-style verbose print instead of `Objects.toString`, because in for a penny.

I don't see any benefit at all in splitting separate commits into separate PRs. Maybe there's benefit in splitting the `-Vcase` flag to a separate commit and the one line in synth methods that uses it.

som-snytt committed:
Clarifying alignments

som-snytt committed:
Honor maxPrintString to limit REPL output

`replStringOf` takes an int that is the max len
to print, not the max number of elements.

Special overloads also take a verbose flag
for printing element names and quotes around
String values.

The 3-param overload does not inject extra
newlines, so that Scala 3 REPL need not strip
them.

som-snytt committed:
Enable verbose toString option for REPL

som-snytt commented:
Rebased and squashed for REPL printing.

SethTisue commented:
@som-snytt what's the status here? in its present form, are you advocating this be merged?

som-snytt commented:
I will re-review.

som-snytt commented:
This is revised PR which does not enable verbose toString for case classes, but as requested, only enables REPL to limit output correctly, and pretty print case classes, which can be turned off with `-Vcase:false`.

Old:
```
val res0: C = C(42,hello, world)
```
New:
```
val res0: C = C(i = 42, s = "hello, world")
```

som-snytt committed:
Multiline string with escaped chars

sideeffffect commented:
Why has this been closed? :crying_cat_face:

som-snytt commented:
This is another minor effort which could also benefit Scala 3 REPL, but low usage of standard REPL may make it not worth it. If it mattered to someone, it would have happened by now.

som-snytt commented:
I needed this today. What a hassle just to get a case class to print something useful.  Moreover, what's the point of providing the minimum possible `toString`?

```
final case class TestFlags(
  defaultClassPath: String,
  runClassPath: String, // class path that is used when running `run` tests (not compiling)
  options: Array[String],
  javacOptions: Array[String]) {
```

SethTisue commented:
> som-snytt reopened this [4 minutes ago]

som-snytt girded his loins, saddled his horse, sharpened his sword, and prepared for battle!

SethTisue commented:
I think this might fall into the category of things where you need to convince the Scala 3 folks first, and if they're game, we can also consider a Scala 2 backport? Does that seem reasonable? (I haven't thought about this in quite a while, just thinking out loud...)

SethTisue commented:
> low usage of standard REPL may make it not worth it

if this is a reference to Ammonite, I don't think the winds are really blowing that strongly Ammonite's way, given all the improvements that have been made to both the Scala 2 and 3 stock REPLs, and given scala-cli's encroachment on Ammonite from the scripting direction. I'm not counting Ammonite out, but I'm also definitely not in the "it doesn't matter, just use Ammonite" camp (and even less in any "everyone uses Ammonite" camp, because I don't think they do)

som-snytt commented:
Scala 3 is already waiting on this PR to support the PR fixing ...

OK, I see the last bit from last year was escaped chars. I guess for multiline, it needs an interpolator to make it look right? I'll have to review the most recent SIP on escape chars in strings.
```
scala> "a\tb"
val res0: String = "a\tb"

scala> println(res0)
a       b

scala> """sometimes
     | what you want
     | \t is eggs
     | for lunch."""
val res2: String =
s"""sometimes
what you want
\\t is eggs
for lunch."""
```

som-snytt commented:
> if this is a reference to Ammonite

The age of heroes is past. This time belongs to the worksheets.

SethTisue commented:
> This time belongs to the worksheets

Damn kids!

som-snytt committed:
Revert using s for multiline

SethTisue commented:
> I think this might fall into the category of things where you need to convince the Scala 3 folks first

And therefore I'm going to close this to get out of the PR queue — we can always reopen later if things move forward on the Scala 3 side.

som-snytt commented:
There is a misunderstanding. Scala 3 is calling into our runtime, and Scala 3 will also benefit when this is merged.

som-snytt commented:
Or it's time to start publishing my fork that does what I need.

SethTisue commented:
Okay. Let's summon a Scala 3 reviewer or two, but before we do that, can you attempt to make the PR description user-facing and country simple about what this does and why?

(To be clear, I believe I am in favor of this change, in my heart at least, and I'm trying to help you give it the best chance of making it through... I hope it doesn't come across as being difficult.)

som-snytt commented:
Scala 3 literally needs at least an ellipsis. Can we spare them three dots?

https://github.com/lampepfl/dotty/pull/16011

som-snytt commented:
While forward-porting the tpolecat test, I realized the improved stringOf is still not merged for Scala 3 repl. How long, Oh Lord?

The message comment is updated to `// return value of "tpolecat.toString" is null`, which sounds like the NPE message.

It would be nice to backport, if there is something to backport it on top of.

som-snytt commented:
Sorry, PR, that I forgot your third birthday.

som-snytt commented:
@SethTisue said:

> I'd cut off my pinky for this. (I mean, like, if it would grow back in a few months.)

https://github.com/scala/bug/issues/3967#issuecomment-292387247

som-snytt commented:
Comparing java record

![image](https://github.com/scala/scala/assets/369425/ef7bbed3-34b4-40ba-991a-dd9b40eb9e29)

som-snytt commented:
Relates to https://github.com/scala/scala/pull/10180

SethTisue commented:
Scala 3 REPL now has pprint-based printing
